temporal-sdk/src/main/java/io/temporal/internal/common/SearchAttributesUtil.java

```diff
@@ -0,0 +1,182 @@
+      } catch (DataConverterException e) {
+        throw new DataConverterException("Cannot serialize key " + item.getKey(), e.getCause());
+      }
+    }
+
+    return serializedMap;
+  }
+
+  public static Map<String, Object> deserializeToObjectMap(SearchAttributes serializedMap) {
+    if (serializedMap == null) {
+      return null;
+    }
+
+    Map<String, Object> deserializedMap = new HashMap<>();
+    for (Entry<String, Payload> attribute : serializedMap.getIndexedFieldsMap().entrySet()) {
+      String key = attribute.getKey();
+      if (!parseSearchAttributes(key, attribute.getValue(), deserializedMap)) {
+        String typeString = attribute.getValue().getMetadataMap().get("type").toStringUtf8();
+        Type javaType = stringToJavaType(typeString);
+        if (javaType == null) {
+          log.error("Error parsing Search Attribute: " + key + " of type " + typeString);
+        } else {
+          deserializedMap.put(
+              key, converter.fromPayload(attribute.getValue(), javaType.getClass(), javaType));
+        }
+      }
+    }
+    return deserializedMap;
+  }
+
+  private static boolean parseSearchAttributes(
+      String key, Payload value, Map<String, Object> deserializedMap) {
+    SearchAttribute searchAttribute;
+    try {
+      searchAttribute = SearchAttribute.valueOf(key);
+    } catch (IllegalArgumentException e) {
+      return false;
+    }
+    switch (searchAttribute) {
+      case BatcherNamespace:
+      case BatcherUser:
+      case BinaryChecksums:
+      case ExecutionStatus:
+      case RunId:
+      case TaskQueue:
+      case TemporalChangeVersion:
+      case WorkflowId:
+      case WorkflowType:
+        deserializedMap.put(key, converter.fromPayload(value, String.class, String.class));
+        break;
+      case CloseTime:
+      case ExecutionTime:
+      case StartTime:
+        deserializedMap.put(
+            key, converter.fromPayload(value, LocalDateTime.class, LocalDateTime.class));
+        break;
+      case ExecutionDuration:
+      case HistoryLength:
+      case StateTransitionCount:
+        deserializedMap.put(key, converter.fromPayload(value, Integer.class, Integer.class));
+        break;
+    }
+    return true;
+  }
+
+  public static Type stringToJavaType(String type) {
+    SearchAttributeType attributeType;
+    try {
+      attributeType = SearchAttributeType.valueOf(type);
+    } catch (IllegalArgumentException e) {
+      return null;
+    }
+    switch (attributeType) {
+      case String:
+      case Keyword:
+        return String.class;
+      case Int:
+        return Integer.class;
+      case Double:
+        return Double.class;
+      case Bool:
+        return Boolean.class;
+      case Datetime:
+        return LocalDateTime.class;
+    }
+    return null;
+  }
+
+  public static String javaTypeToEncodedType(Class type) {
+    String stringName = type.getSimpleName();
+    switch (stringName) {
+      case "String":
```

Comment:
Why strings though? We at least can do something like
String.class.getSimpleName to get a bit more compile-time checks here.
Also, it can be a good idea to use fully qualified names here. There is org.joda.time.LocalDateTime for example.